### ISSUE TITLE
Spack doesn't handle compilers that require LD_LIBRARY_PATH to work

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#332]
I am using `gpc`, and HPC system in Toronto. Spack doesn't correctly detect the system compilers, so I `module add gcc` and then `spack compiler add gcc` to get a working compiler.

Unfortunately, this compiler requires `LD_LIBRARY_PATH`:

```
LD_LIBRARY_PATH=/scinet/gpc/compilers/gcc-4.6.1/requiredlibs/lib:/scinet/gpc/compilers/gcc-4.6.1/lib:/scinet/gpc/compilers/gcc-4.6.1/lib64:/scinet/gpc/toolbin/x11/lib
```

When Spack tries to build OpenMPI, it resets `LD_LIBRARY_PATH`, leading then to linker failures since certain C++ libraries are not found.

**Comment by KineticTheory** [comment]
**What is the recommended work around for this issue?** On LANL HPC, essentially all installed compilers require `LD_LIBRARY_PATH` to be set (usually via a modulefile).  For example, to use `gcc@4.8.2` I need:
`LD_LIBRARY_PATH=/usr/projects/hpcsoft/toss2/common/gcc/4.8.2/lib64:/usr/projects/hpcsoft/toss2/common/gcc/4.8.2/lib`
Without that environment variable set, I see errors like this
`/usr/projects/hpcsoft/toss2/common/gcc/4.8.2/bin/../libexec/gcc/x86_64-unknown-linux-gnu/4.8.2/cc1plus: error while loading shared libraries: libcloog-isl.so.4: cannot open shared object file: No such file or directory`
when trying to build anything from spack.

I don't want spack to install the compiler because I need to use a prebuilt (i.e.: customized) MPI that is associated with a particular installed compiler on each HPC system.

**Comment by adamjstewart** [comment]
#6 also shares the same problem and is the oldest open Issue. We should probably find a way to solve this. My suggestion would be allowing `~/.spack/compilers.yaml` to have a way to set environment variables whenever a compiler is used. This would also be helpful for Intel.

**Comment by hegner** [comment]
I once did things for my own environment. I just re-surrected it and put a commit onto a new feature branch in my fork:
  https://github.com/hegner/spack/commit/1bda8ac33bf6d9b25bda06974b712731851eed9e
This allows you to add `LD_LIBRARY_PATH : <whatever>`  to your compiler config.

If desired, I could make it into a less-hacky PR.

**Comment by adamjstewart** [comment]
I think that would be pretty cool. You would have to make sure that `LD_LIBRARY_PATH` gets unset in `cc` though. Then it would be set if and only if `~/.spack/compilers.yaml` contains that line.

**Comment by hegner** [comment]
@adamjstewart  Opened the PR. Let's continue the discussion there.

**Comment by becker33** [comment]
Would this be resolved by specifying a module to load for each compiler using the modulecmd executable? That module could be loaded while setting up the build environment.

**Comment by tgamblin** [comment]
@becker33: that solves it for most cases but I think this is also necessary and useful for systems where there are no modules or the compilers aren't necessarily in modules.

**Comment by tgamblin** [comment]
@becker33: see #943 too

**Comment by tgamblin** [comment]
Fixed by #2279
